Feature/ci build netgear rax40

rmelotte opened the pull request:
Keeping it as a draft until we can confirm all tests from CI pass. 
Will also need to integrate changes from #616 .

rmelotte committed:
tools: docker: openwrt: allow specifying a build variant

The new version of the prplWrt Makefile defines two variants:
prplmesh-dwpal and prplmesh-nl80211. We cannot use "prplmesh" anymore,
as CONFIG_PACKAGE_prplmesh does not exist anymore, and compiling
package/feeds/iwlwav/prplmesh will compile any default version (which
may not be the one we need).

tools/docker/builder/openwrt/Dockerfile: add a new argument,
  PRPLMESH_VARIANT, to be able to select the correct variant.

tools/docker/builder/openwrt/configs/prplmesh_config: remove the
  prplmesh configuration since we now need to select a specific
  variant.

tools/docker/builder/openwrt/build.sh: add PRPLMESH_VARIANT, and
display PRPLMESH_VARIANT and TARGET with dbg.

Note that this change is backward compatible in the sense that it can
still be used to compile with Makefiles using "prplmesh" alone instead
of adding variants (if PRPLMESH_VARIANT is kept empty, "prplmesh" will
be used).

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tools: docker: openwrt: add support for the intel feed.

Some targets, like the rax40, have dependencies from the intel feed.

Dockerfile:
  - add a new INTEL_FEED argument.
  - when building the SDK, the only thing we need from the feed is the
    intel_mips target, so install it explicitly.
  - when using the SDK, install anything from the intel feed first,
    because it overrides some packages from the "packages" feed.

build.sh:
  - export and use INTEL_FEED if it's set.
  - remove the -f command line option. Allowing the use of many
  different feeds from command line options will be difficult to
  maintain. Also the feeds versions are tied to specific devices at
  the moment, so it makes sense to set the feed version based on the
  target, and not manually.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
ci: build for the netgear rax40

In order to allow building easily for the rax40, add a new target to
the build script.

The rax40 needs the intel feed to build its dependencies. Note that it
uses a commit from a hotfix branch for now, until the required fix is
merged into the main branch.

At the moment, it's not possible to build for both the nl80211-based
targets and the dwpal-based target using the same prpl feed.
To still be able to tests prplMesh with both variants until this is
solved, specify a different PRPL_FEED version for the rax40.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

arnout commented on tools/docker/builder/openwrt/build.sh:
```diff
@@ -26,6 +26,7 @@ build_image() {
            --build-arg OPENWRT_VERSION \
            --build-arg TARGET \
            --build-arg PRPL_FEED \
+	   --build-arg PRPLMESH_VARIANT \
```
Yet another tab problem...

arnout commented on tools/docker/builder/openwrt/build.sh:
```diff
@@ -86,6 +87,8 @@ main() {
     dbg "IMAGE_ONlY=IMAGE_ONLY"
     dbg "TARGET_DEVICE=$TARGET_DEVICE"
     dbg "TAG=$TAG"
+    dbg "TARGET=$TARGET"
```
This is an unrelated change and I'm not sure if it's very relevant.

arnout commented on tools/docker/builder/openwrt/scripts/build.sh:
```diff
@@ -7,5 +7,5 @@ cp -r /home/openwrt/prplMesh_source /home/openwrt/prplMesh
 make package/feeds/prpl/prplmesh/prepare USE_SOURCE_DIR="/home/openwrt/prplMesh" V=s
-make package/feeds/prpl/prplmesh/compile V=sc -j"$(nproc)"
-find bin -name 'prplmesh_*.ipk' -exec cp -v {} "prplmesh-${TARGET}-${OPENWRT_VERSION}-${PRPLMESH_VERSION}.ipk" \;
+make package/feeds/prpl/prplmesh/compile VARIANT="$PRPLMESH_VARIANT" V=sc -j"$(nproc)"
```
This shouldn't be needed since the variant is specified in the config, right? 

Also, since it's an ARG and not an ENV, is it even saved in the environment when running the container?

rmelotte replied:
Last time I tried it didn't use the correct variant even when it was selected in the config file.

This was an attempt to force it to use the correct variant, but it was a bad idea: `VARIANT` will be the same for all the packages, which means that the hostapd variant for example, will be `-dwpal` or `-nl80211`...

Removing.

arnout commented on tools/docker/builder/openwrt/Dockerfile:
```diff
@@ -41,7 +44,12 @@ RUN git clone "$OPENWRT_REPOSITORY" openwrt \
+RUN if [ -n "$INTEL_FEED" ] ; then \
+	echo "src-git intel $INTEL_FEED" >> feeds.conf.default \
+	&& scripts/feeds update intel \
+	&& scripts/feeds install intel_mips ; \
```
Can you explain better in the commit message why it is necessary to explicitly install this one?

rmelotte replied:
Done.

arnout commented on tools/docker/builder/openwrt/build.sh:
```diff
@@ -10,7 +10,6 @@ usage() {
     echo "  options:"
     echo "      -h|--help - show this help menu"
     echo "      -d|--target-device the device to build for"
-    echo "      -f|--prpl-feed the prpl feed to use"
```
Can you extend the help text with the environment variables that affect the build? I.e., PRPL_FEED and INTEL_FEED. And include their defaults as well.

rmelotte replied:
Done.

arnout commented on tools/docker/builder/openwrt/build.sh:
```diff
@@ -64,6 +64,12 @@ main() {
+                    netgear-rax40)
+                        TARGET=intel_mips
+                        PRPLMESH_VARIANT="-dwpal"
+                        PRPL_FEED="https://git.prpl.dev/prplmesh/iwlwav.git^e9ee6b109764dbb91b6cb399ccc0f78e5b04e260"
```
Is this really still needed? Can't we still use this feed by specifying the variant?

I'm running a test build, if successful I'll push that with fixup commits.

rmelotte replied:
When using the latest feed for the nl80211 targets, dwpal gets compiled although it should not (the intel feed is missing so it fails with safeclibs missing).
I suggest we first merge this PR using the 2 separate feed versions and open an issue about it.
That way we can finally build the dwpal variant in CI and at least spot potential compilation issues introduced by other changes early.

arnout reviewed (approved):
Easily fixed comments only.

I did a rebase on master to be able to test some things, so I'll force-push that after running a test build.

arnout commented:
Build didn't complete yet but I had to go home, so let's wait for CI to report if it works or not...

rmelotte commented:
For reference, latest pipeline was here: https://gitlab.com/prpl-foundation/prplMesh/pipelines/106383852
It still fails because it compiles dwpal even for the nl80211 targets.
I'll force push again with the 2 different feeds version, and the changes that were requested. 

I kept a copy of the 2 additional fixup commits in case we still want them.

rmelotte commented:
I changed the value of `INTEL_FEED` to be an empty string (to make it obvious that it's being used in the script, like we do for the other variables). Because of this, the CI needs to rebuild the whole image.

I had to cancel the latest one to avoid compiling the same image twice at the same time (which is why github shows the pipeline as canceled).

Marking it as ready to review already, but we will need to re-run the pipeline once the current build is done.

rmelotte committed:
tools: docker: openwrt: fix some debug messages

In commit 175930d8a1ed045a791aac654abc68b03f57e70e, the loop on the
target was moved outside of the case block, and after the debug
messages.

As a consequence, the debug messages didn't include the
target-specific changes to the variables.

Move the debug messages after the loop, so that the correct values get
printed.

While we're at it, add the missing "INTEL_FEED" variables to the debug
messages.  Also move the message about the container name after
the variable initialization, so that the value gets printed.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tools: docker: openwrt: make sure the container is always removed

The build.sh script uses errexit. Therefore, it exits at the first
encountered error. Because of this, it may not always remove the
containers it creates.

Use 'trap' to remove the containers at exit no matter if the test
succeeded or not.

Also, add the '-f' option to 'docker rm', to force the container to be
removed even if it didn't stop after the 'docker run' invocation.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tools: docker: openwrt: make the container names unique

Since commit 879ecb971e119d81969d73394aa73e90d59de231, the image names
are unique, so that multiple concurrent jobs don't use each other's
images. The container names however, were named with a timestamp, but
weren't necessarily unique. Now that we run multiple jobs in parallel,
it's essential that those container names are also unique.

Change container_name to use uuidgen instead of a timestamp. While
we're at it, append TARGET to the container name, so that we know
which target the container is building for.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte commented:
The build failed because of a container name clash, so I added an additional commit to use unique container names (image names were already unique, containers names were only timestamped).

I also noticed the containers weren't always removed, so I added another commit to make sure containers are removed even when the build fails.

Finally, I also fixed some debug messages which printed wrong values.

rmelotte commented:
Added quotes to the target in gitlab-ci.yml, since every other target are quoted.

rmelotte commented:
Rebased on master.
